proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,257 @@
+accounts, which helps to track and verify the reward distribution progress, will be
+updated.
+
+### `EpochRewardHistory` Sysvar Account
+
+`EpochRewardHistory` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardHistory` is an array of `RewardHistoryEntry`,
+where each entry corresponds to the reward for one particular epoch. The
+maximum depth of the history is chosen to be the same as the depth of stake
+history for consistency, which is 512.
+
+The layout of `EpochRewardHistory` sysvar is shown in the following pseudo code.
+
+```
+struct RewardHistoryEntry {
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   root_hash: Option<Hash>,               // hash computed from all EpochRewardReserves
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+The `EpochRewardHistory` is updated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch
+rewards and vote account rewards become available at this time. Once the vote
+account rewards have been distributed, a `RewardHistoryEntry` will be added to
+the `EpochRewardHistory` array with the total amount of reward, a distributed
+value equal to the sum of the vote account rewards. If the length of the history
+array exceeds the maximum depth (512), the oldest entries will be removed from
+the array (i.e., "popped").
+
+The `root_hash` is the hash of all the `EpochRewardReserve' accounts' hashes. Initially,
+at the start of the epoch's first block, the `root_hash` is set to None. Once all
+the `EpochRewardReserve` accounts are filled and the reward computation phase
+is complete, the `root_hash` is calculated by hashing all the
+`EpochRewardReserve` hashes.
+
+```
+root_hash = Hash(EpochRewardReserve[..].rewards_hash)
+
+
+
+                     +-------------+
+                     |             |
+                     |  root_hash  |
+                     |             |
+                     +------+------+
+                            |
+    +--------------+--------+-----------+
+    |              |                    |
++---+----+   +-----+--+             +---+-----+
+|reserve |   | reserve|   ......    | reserve |
+| hash   |   |  hash  |             |  hash   |
++--------+   +--------+             +---------+
+
+```
+
+### `EpochRewardReserve` Sysvar Account
+
+The address for a particular `EpochRewardReserve` account is derived from the
+base Sysvar RewardReserve address, by hashing
+`SysvarRewardReserve11111111111111111111111111111111` with the u64 block height
+for the corresponding reward distribution block during the reward
+distribution phase.
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve11111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   rewards_hash: Hash,
```

Comment:
little nit, to be consistent with the new `root_hash` diagram
```suggestion
   reserve_hash: Hash,
```

If this looks like there's a couple other instances of `reserve_hash` in the doc that could be renamed as well